### PR TITLE
downloadable tx summary and logs

### PR DESCRIPTION
**Pull request by josepfo**

### CODE REVIEW

**Comment by josepfo** [comment]
[Implement downloadable transaction summary / debug logs](https://trello.com/c/rr6aGoW7/641-implement-downloadable-transaction-summary-debug-logs)

File: lib/new-admin/services/transactions.js [hunk 1]
```diff
@@ -162,4 +162,33 @@ function cancel (txId) {
...
+  txs.*,
+  (extract(epoch from (now() - greatest(txs.created, txs.confirmed_at))) * 1000) >= $2 as expired
+  from cash_out_txs txs
+  where txs.id=$1`
+
+  if (txClass === 'cashIn') {
+    return db.oneOrNone(cashInSql, [cashInTx.PENDING_INTERVAL, txId])
+  } else {
+    return db.oneOrNone(cashOutSql, [txId, REDEEMABLE_AGE])
+  }
```
Comment (chaotixkilla) [inline lib/new-admin/services/transactions.js]:
```suggestion
  return txClass === 'cashIn'
    ? db.oneOrNone(cashInSql, [cashInTx.PENDING_INTERVAL, txId])
    : db.oneOrNone(cashOutSql, [txId, REDEEMABLE_AGE])
```

File: lib/new-admin/services/transactions.js [hunk 2]
```diff
@@ -162,4 +162,33 @@ function cancel (txId) {
...
+}
+
+function getTxAssociatedData (txId, txClass) {
+  const billsSql = `select 'bills' as bills, b.* from bills b where cash_in_txs_id = $1`
+  const actionsSql = `select 'cash_out_actions' as cash_out_actions, actions.* from cash_out_actions actions where tx_id = $1`
+  if (txClass === 'cashIn') {
+    return db.manyOrNone(billsSql, [txId])
+  } else {
+    return db.manyOrNone(actionsSql, [txId])
+  }
```
Comment (chaotixkilla) [inline lib/new-admin/services/transactions.js]:
```suggestion
  return txClass === 'cashIn'
    ? db.manyOrNone(billsSql, [txId])
    : db.manyOrNone(actionsSql, [txId])
```

**Review by chaotixkilla — CHANGES_REQUESTED** [review]

**Review by chaotixkilla — APPROVED** [review]

File: new-lamassu-admin/src/pages/Transactions/DetailsCard.js [hunk 3]
```diff
@@ -55,6 +84,30 @@ const DetailsRow = ({ it: tx }) => {
...
+    fetchSummary({ variables: { txId, from, until, deviceId, txClass } })
+  }
+
+  const createCsv = logs => {
+    const zipFilename = `tx_${tx.id}_summary.zip`
+    const filesNames = R.keys(logs)
+    R.map(name => zip.file(name + '.csv', logs[name]), filesNames)
+    zip.generateAsync({ type: 'blob' }).then(function(content) {
+      FileSaver.saveAs(content, zipFilename)
+    })
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Transactions/DetailsCard.js]: You can use async await here.

```js
const content = await zip.generateAsync({ type: 'blob' })
FileSaver.saveAs(content, zipFilename)
```

File: new-lamassu-admin/src/pages/Transactions/DetailsCard.js [hunk 4]
```diff
@@ -206,6 +259,22 @@ const DetailsRow = ({ it: tx }) => {
             errorElements
           )}
         </div>
+        <div>
+          {
+            <div
+              onClick={() => {
+                loadAndTxSummary(tx)
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Transactions/DetailsCard.js]: This function name doesn't really tell me anything. Why not `downloadRawLogs`

File: new-lamassu-admin/src/pages/Transactions/DetailsCard.js [hunk 5]
```diff
@@ -55,6 +84,30 @@ const DetailsRow = ({ it: tx }) => {
...
+  const from = moment(tx.created)
+    .subtract(MINUTES_OFFSET, 'm')
+    .format()
+  const until = moment(tx.created)
+    .add(MINUTES_OFFSET, 'm')
+    .format()
+
+  const [fetchSummary] = useLazyQuery(TX_SUMMARY, {
+    onCompleted: data => createCsv(data)
+  })
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Transactions/DetailsCard.js]: move this to the top of the function, alongside any other hooks we call

**Review by RafaelTaranto — CHANGES_REQUESTED** [review]

File: new-lamassu-admin/src/pages/Transactions/DetailsCard.js (hunk 4 above)
Comment (josepfo) [reply new-lamassu-admin/src/pages/Transactions/DetailsCard.js]: It was supposed to be loadTxSummary but your suggestion makes more sense

File: lib/new-admin/graphql/types/transaction.type.js [hunk 6]
```diff
@@ -48,6 +48,8 @@ const typeDef = gql`
   type Query {
     transactions(from: Date, until: Date, limit: Int, offset: Int, deviceId: ID): [Transaction]
     transactionsCsv(from: Date, until: Date, limit: Int, offset: Int): String
+    transactionCsv(id: ID, txClass: String): String
+    txAssociatedDataCsv(id: ID, txClass: String): String
```
Comment (RafaelTaranto) [inline lib/new-admin/graphql/types/transaction.type.js]: This needs `@auth`, rebase to see it working.

**Review by RafaelTaranto — CHANGES_REQUESTED** [review]

File: new-lamassu-admin/src/pages/Transactions/DetailsCard.js [hunk 7]
```diff
@@ -206,6 +258,22 @@ const DetailsRow = ({ it: tx }) => {
...
+            <div
+              onClick={() => {
+                downloadRawLogs(tx)
+              }}>
+              <Label>Other actions</Label>
+              <ActionButton
+                color="primary"
+                Icon={Download}
+                className={classes.downloadRawLogs}>
+                {'Download raw logs'}
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Transactions/DetailsCard.js]: I don't think you need the brackets nor the quotes here.

File: new-lamassu-admin/src/pages/Transactions/DetailsCard.js [hunk 8]
```diff
@@ -206,6 +258,22 @@ const DetailsRow = ({ it: tx }) => {
             errorElements
           )}
         </div>
+        <div>
+          {
+            <div
+              onClick={() => {
+                downloadRawLogs(tx)
+              }}>
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Transactions/DetailsCard.js]:
```suggestion
              onClick={() => downloadRawLogs(tx)}
```

**Review by RafaelTaranto — CHANGES_REQUESTED** [review]

**Review by RafaelTaranto — APPROVED** [review]

**Pull request merged**